### DISCUSSION TITLE
Typer not compatible with uppercase lettered arguments.

### DISCUSSION THREAD

**Discussion by thisiscam** [fastapi/typer#1556] [Questions] [answered]
### Describe the bug

Typer does not seem to be compatible with uppercase lettered arguments.

### To Reproduce

Steps to reproduce the behavior with a minimum self-contained file.

Replace each part with your own scenario:

* Create a file `main.py` with:

```Python
import typer

app = typer.Typer()


@app.command()
def f(var_with_UPPERCASE: int, lowercase_var: int):
  return var_with_UPPERCASE + lowercase_var


app()
```

* Call it with:

```bash
python main.py 1 2
```

* It outputs:

```
TypeError: f() got an unexpected keyword argument 'var_with_uppercase'
```

* But I expected it to output:

```
3
```

### Expected behavior

Typer should preserve the variable case.

### Additional context

I suspect this is related to Typer's name mangling with underscores, which always turns the variable to lower case? I think a more robust solution is to allow any case + "_"/"-" combinations, or at least an option in `Typer(...)` to enable this feature will be great!

**Comment by aakuyr** [comment]
The issue seems to stem not from Typer but Click itself which doesn't seem to preserve the casing of the options/method arguments.  
[Here is the click issue for the same thing I found](https://github.com/pallets/click/issues/837)
Since Typer seems to inherit arguments and command classes from click this has most likely carried over. The reason for this is lower case with dashes is friendlier for CLI use cases.

**Comment by tiangolo** [comment, marked as answer]
Thanks for the help @aakuyr! :bow:

Yep, it comes from Click, and as Click is what does the work underneath, that is handled there.

Also have in mind that for many scenarios letter case doesn't have significance in Windows, having something that requires uppercase might be confusing, or it could work on some platforms but not on others.

**Comment by thisiscam** [comment]
Thanks. That kind of makes sense. 

I forgot what exact scenario I was under but I definitely had a case where uppercase letter is preferred. 
If I have to guess, sometimes I use uppercase symbols that have standard interpretations in math (changing to lower case alters that meaning). 

If this is not going to be supported, maybe raise a more interpretable error explicitly marking uppercase as not allowed?

**Comment by jplusc** [comment]
I know this ticket is ~4 yrs old, and i've been using Typer for 3yrs of that, but I just today struggled for way to long to figure out why my optional argument wasn't working.

I kept double checking the syntax for an optional string argument like:
```
def derecode(additionalComments: Annotated[Optional[str], typer.Argument()] = None):
def derecode(additionalComments: str = typer.Argument(default=None)):
```

and kept getting the error `TypeError: derecode() got an unexpected keyword argument 'additionalcomments'`

I see now it is a design decision of Click since it is awkward to type mixed case at a cli, but for an optional argument, it is never actually typed, at this point it is just a variable name.

I kinda agree with above, that maybe a raise a more interpretable error should be raised.  or at least maybe point this out in the docs.  I remember seeing the section on dashes and underscores, but I don't recall anything about mixed case.
Maybe the limitation can also be highlighted in the section about optional args?



Anyway, mainly adding this comment so others hitting the unexpected keyword error might find this thread.

*thanks for all you do!*

**Comment by YuriiMotov** [comment]
As I see it, this issue has two sides:
* Typer\Click don't preserve letter case in names (this is intended as described in https://github.com/pallets/click/issues/837)
* Typer\Click fails to parse the argument if the name of variable contains uppercase letters

In the case `python main.py 1 2` the argument name is not used, but it still doesn't work. I would expect that `1` parsed as first argument value and placed into the `var_with_UPPERCASE` variable.

At the same time variable names like `var_with_UPPERCASE` go against the Python's naming conventions.

Considering low level of support and assuming this behavior comes from Click, I think we can close this issue.
